Don't clamp color to [0, 1] in Linear tonemapping

NeoSpark314 opened the pull request:
At the moment there is no possibility to render a 3d scene to an HDR viewport texture without the color being clamped to the range [0.0, 1.0] (the behaviou of the "Linear" tone mapping operator is to just clamp (see https://github.com/godotengine/godot/blob/79a480a55e1ebada7f2987afeeb2039a39c8666b/drivers/gles3/shaders/tonemap.glsl#L180).
 
I ran into this issue when exporting a 3d scene rendered to a viewport texture to an .exr file with the need to retain values > 1.0. 

One potential solution is to disable tonemapping completely via the here proposed tonemapping option "None".

I think this option could also fix https://github.com/godotengine/godot/issues/26821 as an alternative solution.

akien-mga commented on doc/classes/Environment.xml:
```diff
@@ -305,7 +305,7 @@
 		</constant>
 		<constant name="TONE_MAPPER_LINEAR" value="0" enum="ToneMapper">
-			Linear tonemapper operator. Reads the linear data and performs an exposure adjustment.
+			Linear tonemapper operator. Reads the linear data and performs an exposure adjustment by clamping to [0.0,1.0].
```
Style wise, there should be a space after the comma in `[0.0, 1.0]`. Additionally, the parser will fail here as `[` and `]` are used for BBcode-like tags, so you should enclose it in `[code][0.0, 1.0][/code]`.

NeoSpark314 replied:
Yes; thanks; changed and pushed it to the pull request.

clayjohn reviewed (approved):
Looks good to me! But we should check with reduz that this is a desired feature before merging.

akien-mga commented:
Some comments from IRC:
```
18:21 <Akien> reduz: talking about tonemapping, ok with #31571 ?
18:21 <IssueBot> #31571: Added an additional tonemapping option called 'none'  | https://git.io/fjNYv
18:37 <reduz> Akien: if you disable tonemapping, you get the same effect, so not quite sure what he wants to do
18:38 <Akien> reduz: they want color values unclamped for hdr it seems.
18:38 <Akien> > 1.0
18:39 <reduz> ah, I am not sure if that is the best way to do it tbh
18:40 <reduz> we could probably just not clamp it and thats it
18:40 <reduz> i personally dont see much reason to clamp it per se
18:41 <reduz> in fact, i think we shouldnt do it
18:56 <Akien> I guess they'd be fine with that.
19:02 <reduz> Akien: let me take the chance to change this in the 4.0 branch, since I am working exactly on this code right now
19:02 <reduz> so worst case we backport it
19:04 <Akien> reduz: ok
19:46 <clayjohn> reduz: if you overhaul the tonemapping code in Vulkan, I can backport to 3.2. I'll just do GLES2 and GLES3 at the same time.
19:56 <reduz> clayjohn: I am mostly cleaning it up I guess
19:57 <clayjohn> reduz: that makes sense. I should wait then cause it should be easy enough to port.
```

NeoSpark314 commented:
Thanks for the discussion. Not clamping at all for the current "linear" case would solve the problem as well and would be equally good from my point of view. Adding the additional 'none' option would only be needed for backwardscompatibility of rendering clamped into hdr viewport textures.

clayjohn reviewed (changes requested):
Thanks for your work so far! After discussing with reduz, he has confirmed that his preference is to just not clamp when using ``linear`` instead of providing a new option. Since you have already put in the work, do you mind updating this PR to just return ``color`` when ``linear`` is used instead of clamped color?

NeoSpark314 commented:
Yes; I'll change the PR.

NeoSpark314 committed:
Removed clamping of the Linear tonemapping when KEEP_3D_LINEAR

Changed the behaviour of the Linear tonemapping operator to not clamp to [0, 1] range
in the case when KEEP_3D_LINEAR is defined. This allows to render values > 1.0 in
floating point texture targets (via Viewport) for further processing or saving high
dynamic range data into files. This only works when no color conversion is active.

NeoSpark314 commented:
I made the changes...
But I looked a bit closer what this means. All optional post processing steps (and the sRGB conversion) that come after the tonemapping call assume clamped values. I don't think it is good to mess this up with the Linear operator. My proposed solution to this in the PR is to now clamp all values in the respective post processing steps to [0, 1]. This still allows to render to float render targets values > 1.0 when all postprocessing is disabled and sRGB conversion is turned off. But looks not very nice in the code :(.

NeoSpark314 commented:
Right; silently clamping might lead to confusion. And in vulkan most of the clamping is not done.... I can adapt my PR to reflect this behavior
but I think it could lead to a lot of unexpected behavior with existing setups as Linear is the default (but I need to create some test scenes to confirm this). Materials with emission or bright light sources may quickly go over 1.0.

I don't know with the BCS computation but at least in the case of `linear_to_srgb(color)` this would result in weird values (the comment already says that values have to be between [0, 1]). Also color correction assumes a lookup in [0, 1] in the lookup texture.
(and in the vulkan code two glow modes silently clamp.)

Do you think it is enough to then document that for postprocessing to work propperly a Tonemapper other then linear has to be chosen?

NeoSpark314 commented:
Yes; this sounds like a good idea to clamp before the linear_to_srgb conversion. This will keep the old behaviour for most exisiting cases since keep_3d_linear is off by default (and probably always off for the main viewport). I'll adapt the PR to reflect the Vulkan code now.

clayjohn reviewed (approved):
Looks good to me! Thanks for being so flexible. 

Let's get reduz to take one last look before getting this merged.

akien-mga commented:
Just a style note in case further edits are made: Git commits should be formatted as a short title on the first line, and the rest goes in the body of the commit log (line 3 and beyond).

The current commit log appears wrapped on GitHub, and would appear very long in `git shortlog` compared to all other commit titles kept around 50-70 chars (50 is recommended max, but pretty hard to fit in in my experience). See https://github.com/godotengine/godot/blob/master/CONTRIBUTING.md#format-your-commit-logs-with-readability-in-mind

NeoSpark314 commented:
Yes; I updated the commit message to follow the guidelines better.

akien-mga commented:
Thanks!
